message.d.ts

```diff
@@ -0,0 +1,588 @@
+ * and call it the intuitive way.
+ */
+export declare class Message {
+    /**
+     * Base class for all JsPb messages.
+     *
+     * Several common methods (toObject, serializeBinary, in particular) are not
+     * defined on the prototype to encourage code patterns that minimize code bloat
+     * due to otherwise unused code on all protos contained in the project.
+     *
+     * If you want to call these methods on a generic message, either
+     * pass in your instance of method as a parameter:
+     * someFunction(instanceOfKnownProto,
+     * KnownProtoClass.prototype.serializeBinary);
+     * or use a lambda that knows the type:
+     * someFunction(()=>instanceOfKnownProto.serializeBinary());
+     * or, if you don't care about code size, just suppress the
+     * WARNING - Property serializeBinary never defined on Message
+     * and call it the intuitive way.
+     */
+    constructor();
+    private noStructuralTyping____protobuf_javascript_message_Message;
+    static GENERATE_TO_OBJECT: boolean;
+    static GENERATE_FROM_OBJECT: boolean;
+    static GENERATE_TO_STRING: boolean;
+    static ASSUME_LOCAL_ARRAYS: boolean;
+    /**
+     * Repeated fields numbers.
+     */
+    protected repeatedFields: number[] | undefined | null;
+    /**
+     * Returns the JsPb message_id of this proto.
+     * @return the message id or undefined if this message has no id.
+     */
+    getJsPbMessageId(): string | undefined;
+    /**
+     * Initializes a JsPb Message.
+     * @param msg The JsPb proto to modify.
+     * @param data An initial data array.
+     * @param messageId For response messages, the message id or '' if no message id is specified. For non-response messages, 0.
+     * @param suggestedPivot The field number at which to start putting fields into the extension object. This is only used if data does not contain an extension object already. -1 if no extension object is required for this message type.
+     * @param repeatedFields The message's repeated fields.
+     * @param opt_oneofFields The fields belonging to each of the message's oneof unions.
+     */
+    static initialize(msg: Message, data: any[] | undefined | null, messageId: string | number, suggestedPivot: number, repeatedFields: number[] | null, opt_oneofFields?: number[][] | null): void;
+    /**
+     * Converts a JsPb repeated message field into an object list.
+     * @param field The repeated message field to be converted.
+     * @param toObjectFn The toObject function for this field.  We need to pass this for effective dead code removal.
+     * @param opt_includeInstance Whether to include the JSPB instance for transitional soy proto support: http://goto/soy-param-migration
+     * @return An array of converted message objects.
+     */
+    static toObjectList<T = any>(field: (T | null)[], toObjectFn: ((() => (object | null)) | null) | (() => (object | null)) | null, opt_includeInstance?: boolean): (object | null)[];
```

Comment:
`toObjectFn: (this: T, includeInstance: boolean | undefined, msg: T | null) => (object | null)`

Reply:
Keeping the signature vague is fine since this is internal, but the redundancy is definitely confusing. i.e. the current type is equivalent to just `(() => (object | null)) | null` but "phrased" as `(T) | T` if we let `T` be the simpler type. Maybe not worth spending a lot of effort to reduce the redundancy, but it would help clarity if possible. (I think this would address the redundant `| null | null` comment as well, so resolving that one to remove duplicate suggestions.)